samples/bluetooth/bap_broadcast_sink/src/main.c

```diff
@@ -732,89 +682,107 @@ static bool find_valid_bis_in_subgroup_bis(const struct bt_bap_base_subgroup *su
+	uint32_t bis_indexes = 0;
 	int err;
 
-	/* We only want indexes from a single subgroup, so reset between each of them*/
-	*bis_indexes = 0U;
-
 	err = bt_bap_base_subgroup_codec_to_codec_cfg(subgroup, &codec_cfg);
 	if (err != 0) {
 		printk("Could not get codec configuration: %d\n", err);
 
-		return true;
+		return true; /* continue to next subgroup */
 	}
 
-	err = bt_audio_codec_cfg_get_chan_allocation(&codec_cfg, &chan_allocation, false);
+	/* Get channel allocation at subgroup level */
+	err = bt_audio_codec_cfg_get_chan_allocation(&codec_cfg, &subgroup_chan_allocation, false);
+	if (err == 0) {
+		printk("Channel allocation (subgroup level) 0x%x\n", subgroup_chan_allocation);
+		subgroup_chan_allocation_available = true;
+	} else if (err == -ENODATA) {
+		printk("No channel allocation data in subgroup\n");
+	} else {
+		/* subgroup error */
+		return true; /* continue to next subgroup */
+	}
+
+	struct bis_channel_allocation_data data;
+
+	/* Keep base code id (used in bt_audio_codec_cfg_get_chan_allocation at bis level) */
+	data.bap_base_codec_id.id = codec_cfg.id;
+	data.bap_base_codec_id.cid = codec_cfg.cid;
+	data.bap_base_codec_id.vid = codec_cfg.vid;
+	data.cnt = 0;
+
+	/* Get channel allocation at BIS level */
+	err = bt_bap_base_subgroup_foreach_bis(subgroup, bis_get_channel_allocation_cb, &data);
 	if (err != 0) {
-		printk("Could not find subgroup channel allocation: %d - Looking in the BISes\n",
-		       err);
+		printk("Get channel allocation error %d\n", err);
 
-		/* Find chan alloc in BIS */
-		if (find_valid_bis_in_subgroup_bis(subgroup, bis_indexes)) {
-			/* Found BISes with correct channel allocation */
-			return false;
+		return true; /* continue to next subgroup */
+	}
+
+	/* If no BIS channel allocation available use subgroup channel allocation instead if
+	 * exists (otherwise mono assumed)
+	 */
+	for (uint8_t i = 0U; i < data.cnt; i++) {
+		if (!data.bis[i].chan_allocation_available) {
+			data.bis[i].chan_allocation = subgroup_chan_allocation_available
+							      ? subgroup_chan_allocation
+							      : BT_AUDIO_LOCATION_MONO_AUDIO;
 		}
-	} else {
-		/* If the subgroup contains a single channel, then we just grab the first BIS index
-		 */
-		if (bt_audio_get_chan_count(chan_allocation) == 1 &&
-		    chan_allocation == CONFIG_TARGET_BROADCAST_CHANNEL) {
-			uint32_t subgroup_bis_indexes;
+	}
 
-			/* Set bis_indexes to the first bit set */
-			err = bt_bap_base_subgroup_get_bis_indexes(subgroup, &subgroup_bis_indexes);
-			if (err != 0) {
-				/* Should never happen as that would indicate an invalid
-				 * subgroup If it does, we just parse the next subgroup
-				 */
-				return true;
-			}
+	/* Get the BIS indexes */
+	for (uint8_t i = 0U; i < data.cnt; i++) {
+		if (data.bis[i].chan_allocation == CONFIG_TARGET_BROADCAST_CHANNEL ||
+		    (data.bis[i].chan_allocation & CONFIG_TARGET_BROADCAST_CHANNEL) != 0) {
+			bis_indexes |= BT_ISO_BIS_INDEX_BIT(data.bis[i].index);
+		}
+	}
```

Comment:
If `CONFIG_TARGET_BROADCAST_CHANNEL` == LEFT | RIGHT and `data.bis[0].chan_allocation` is LEFT, `data.bis[1].chan_allocation` is RIGHT and `data.bis[2].chan_allocation` is LEFT | RIGHT then `bis_indexes` becomes `BIT(0) | BIT(1) | BIT(2)` which is not what we want. In this case we want only either `BIT(0) | BIT(1)` or `BIT(2)`.